File: sdk/go/billing.go
package mashgate

import (
	"context"
	"fmt"
)

// ────────────────────────────────────────────────────────────────────────────
// BillingClient — platform subscription, plan, invoice, payment method.
// Mirrors BillingService in contracts/proto/v1/billing.proto (29 RPCs).
//
// Usage:
//
//	sub, err := client.Billing.GetSubscription(ctx)
//	plans, err := client.Billing.ListPlans(ctx)
// ────────────────────────────────────────────────────────────────────────────

// BillingClient wraps billing REST endpoints behind the main mashgate.Client.
type BillingClient struct {
	c *Client
}

// ListPlans returns all platform plans available to the current tenant.
func (b *BillingClient) ListPlans(ctx context.Context) ([]*BillingPlan, error) {
	var out struct {
		Plans []*BillingPlan `json:"plans"`
	}
	if err := b.c.do(ctx, "GET", "/v1/billing/plans", nil, &out); err != nil {
		return nil, err
	}
	return out.Plans, nil
}

// GetPlan retrieves a platform plan by ID.
func (b *BillingClient) GetPlan(ctx context.Context, planID string) (*BillingPlan, error) {
	var out BillingPlan
	if err := b.c.do(ctx, "GET", "/v1/billing/plans/"+planID, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetSubscription returns the current tenant's active subscription.
func (b *BillingClient) GetSubscription(ctx context.Context) (*BillingSubscription, error) {
	var out BillingSubscription
	if err := b.c.do(ctx, "GET", "/v1/billing/subscription", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ChangePlan switches the tenant's subscription to a new plan.
func (b *BillingClient) ChangePlan(ctx context.Context, req ChangePlanRequest) (*BillingSubscription, error) {
	var out BillingSubscription
	if err := b.c.do(ctx, "POST", "/v1/billing/subscription/change-plan", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// CancelPlan cancels the current subscription (effective at period end unless
// CancelImmediately is set).
func (b *BillingClient) CancelPlan(ctx context.Context, req CancelPlanRequest) (*BillingSubscription, error) {
	var out BillingSubscription
	if err := b.c.do(ctx, "POST", "/v1/billing/subscription/cancel", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// PreviewPlanChange computes prorated charges + effective dates for a hypothetical
// plan switch without applying it. Useful for confirmation UIs.
func (b *BillingClient) PreviewPlanChange(ctx context.Context, req PreviewPlanChangeRequest) (*PreviewPlanChangeResponse, error) {
	var out PreviewPlanChangeResponse
	if err := b.c.do(ctx, "POST", "/v1/billing/subscription/preview-change", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListPaymentMethods returns all billing payment methods registered for the tenant.
func (b *BillingClient) ListPaymentMethods(ctx context.Context) ([]*BillingPaymentMethod, error) {
	var out struct {
		Methods []*BillingPaymentMethod `json:"methods"`
	}
	if err := b.c.do(ctx, "GET", "/v1/billing/payment-methods", nil, &out); err != nil {
		return nil, err
	}
	return out.Methods, nil
}

// AddPaymentMethod registers a new payment method for the tenant.
func (b *BillingClient) AddPaymentMethod(ctx context.Context, req AddBillingPaymentMethodRequest) (*BillingPaymentMethod, error) {
	var out BillingPaymentMethod
	if err := b.c.do(ctx, "POST", "/v1/billing/payment-methods", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SetDefaultPaymentMethod marks a method as the auto-billing default.
func (b *BillingClient) SetDefaultPaymentMethod(ctx context.Context, methodID string) (*BillingPaymentMethod, error) {
	var out BillingPaymentMethod
	path := fmt.Sprintf("/v1/billing/payment-methods/%s/default", methodID)
	if err := b.c.do(ctx, "POST", path, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RemovePaymentMethod deletes a billing payment method.
func (b *BillingClient) RemovePaymentMethod(ctx context.Context, methodID string) error {
	return b.c.do(ctx, "DELETE", "/v1/billing/payment-methods/"+methodID, nil, nil)
}

// ListInvoices returns all billing invoices for the tenant.
func (b *BillingClient) ListInvoices(ctx context.Context) ([]*BillingInvoice, error) {
	var out struct {
		Invoices []*BillingInvoice `json:"invoices"`
	}
	if err := b.c.do(ctx, "GET", "/v1/billing/invoices", nil, &out); err != nil {
		return nil, err
	}
	return out.Invoices, nil
}

// GetInvoice retrieves a single invoice by ID.
func (b *BillingClient) GetInvoice(ctx context.Context, invoiceID string) (*BillingInvoice, error) {
	var out BillingInvoice
	if err := b.c.do(ctx, "GET", "/v1/billing/invoices/"+invoiceID, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// PayInvoice triggers an immediate payment attempt on an open invoice.
func (b *BillingClient) PayInvoice(ctx context.Context, invoiceID string) (*BillingInvoice, error) {
	var out BillingInvoice
	if err := b.c.do(ctx, "POST", "/v1/billing/invoices/"+invoiceID+"/pay", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetCreditBalance returns the tenant's current credit balance.
func (b *BillingClient) GetCreditBalance(ctx context.Context) (*CreditBalance, error) {
	var out CreditBalance
	if err := b.c.do(ctx, "GET", "/v1/billing/credit-balance", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RedeemPromoCode applies a promo code to the tenant.
func (b *BillingClient) RedeemPromoCode(ctx context.Context, code string) (*RedeemPromoCodeResponse, error) {
	var out RedeemPromoCodeResponse
	req := struct {
		Code string `json:"code"`
	}{Code: code}
	if err := b.c.do(ctx, "POST", "/v1/billing/promo/redeem", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdk/python/mashgate/resources/billing.py
"""Billing resource.

Mirrors ``BillingService`` from ``contracts/proto/v1/billing.proto`` —
platform subscription, plans, invoices, payment methods, and credits for
the current tenant. Auth is tenant-scoped; pass an admin JWT or
service-account API key on the parent client.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mashgate.client import MashgateClient


class BillingResource:
    def __init__(self, client: MashgateClient) -> None:
        self._c = client

    # ── Plans ─────────────────────────────────────────────────────────

    def list_plans(self) -> dict[str, Any]:
        """Return all platform plans available to the current tenant."""
        return self._c.request("GET", "/v1/billing/plans")

    def get_plan(self, plan_id: str) -> dict[str, Any]:
        """Retrieve a platform plan by ID."""
        return self._c.request("GET", f"/v1/billing/plans/{plan_id}")

    # ── Subscription ──────────────────────────────────────────────────

    def get_subscription(self) -> dict[str, Any]:
        """Return the current tenant's active subscription."""
        return self._c.request("GET", "/v1/billing/subscription")

    def change_plan(self, *, plan_id: str, immediate: bool | None = None) -> dict[str, Any]:
        """Switch the tenant's subscription to a new plan."""
        body: dict[str, Any] = {"planId": plan_id}
        if immediate is not None:
            body["immediate"] = immediate
        return self._c.request("POST", "/v1/billing/subscription/change", body=body)

    def cancel_plan(
        self,
        *,
        reason: str | None = None,
        immediate: bool | None = None,
    ) -> dict[str, Any]:
        """Cancel the current subscription.

        Effective at period end unless ``immediate`` is set.
        """
        body: dict[str, Any] = {}
        if reason is not None:
            body["reason"] = reason
        if immediate is not None:
            body["immediate"] = immediate
        return self._c.request("POST", "/v1/billing/subscription/cancel", body=body or None)

    def preview_plan_change(self, *, plan_id: str) -> dict[str, Any]:
        """Compute prorated charges + effective dates for a hypothetical plan
        switch without applying it. Useful for confirmation UIs."""
        return self._c.request(
            "POST", "/v1/billing/subscription/preview", body={"planId": plan_id}
        )

    # ── Payment methods ───────────────────────────────────────────────

    def list_payment_methods(self) -> dict[str, Any]:
        """Return all billing payment methods registered for the tenant."""
        return self._c.request("GET", "/v1/billing/payment-methods")

    def add_payment_method(
        self,
        *,
        token: str,
        provider: str | None = None,
        brand: str | None = None,
        last4: str | None = None,
        exp_month: int | None = None,
        exp_year: int | None = None,
        set_default: bool | None = None,
    ) -> dict[str, Any]:
        """Register a new payment method for the tenant."""
        body: dict[str, Any] = {"token": token}
        if provider is not None:
            body["provider"] = provider
        if brand is not None:
            body["brand"] = brand
        if last4 is not None:
            body["last4"] = last4
        if exp_month is not None:
            body["expMonth"] = exp_month
        if exp_year is not None:
            body["expYear"] = exp_year
        if set_default is not None:
            body["setDefault"] = set_default
        return self._c.request("POST", "/v1/billing/payment-methods", body=body)

    def set_default_payment_method(self, method_id: str) -> dict[str, Any]:
        """Mark a method as the auto-billing default."""
        return self._c.request(
            "POST", f"/v1/billing/payment-methods/{method_id}/default"
        )

    def remove_payment_method(self, method_id: str) -> dict[str, Any]:
        """Delete a billing payment method."""
        return self._c.request("DELETE", f"/v1/billing/payment-methods/{method_id}")

    # ── Invoices ──────────────────────────────────────────────────────

    def list_invoices(self) -> dict[str, Any]:
        """Return all billing invoices for the tenant."""
        return self._c.request("GET", "/v1/billing/invoices")

    def get_invoice(self, invoice_id: str) -> dict[str, Any]:
        """Retrieve a single billing invoice by ID."""
        return self._c.request("GET", f"/v1/billing/invoices/{invoice_id}")

    def pay_invoice(self, invoice_id: str) -> dict[str, Any]:
        """Trigger an immediate payment attempt on an open invoice."""
        return self._c.request("POST", f"/v1/billing/invoices/{invoice_id}/pay")

    # ── Credits ───────────────────────────────────────────────────────

    def get_credit_balance(self) -> dict[str, Any]:
        """Return the tenant's current credit balance."""
        return self._c.request("GET", "/v1/billing/credits")

    def redeem_promo_code(self, code: str) -> dict[str, Any]:
        """Apply a promo code to the tenant."""
        return self._c.request("POST", "/v1/billing/credits/redeem", body={"code": code})
